### PR TITLE
[SPARK-44811][BUILD] Upgrade Guava to 33.2.1-jre

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

This PR upgrades Spark's built-in Guava from 14 to 33.2.1-jre

Currently, Spark uses Guava 14 because the previous built-in Hive 2.3.9 is incompatible with new Guava versions. HIVE-27560 (https://github.com/apache/hive/pull/4542) makes Hive 2.3.10 compatible with Guava 14+ (thanks to @LuciferYang)

### Why are the changes needed?

It's a long-standing issue, see prior discussions at https://github.com/apache/spark/pull/35584, https://github.com/apache/spark/pull/36231, and https://github.com/apache/spark/pull/33989

### Does this PR introduce _any_ user-facing change?

Yes, some user-faced error messages changed.

### How was this patch tested?

GA passed.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -343,6 +343,15 @@
         <enabled>false</enabled>
       </snapshots>
     </repository>
+
+    <repository>
+      <id>mvn-repo</id>
+      <url>https://raw.github.com/pan3793/mvn-repo/master</url>
```
Comment (pan3793) [inline pom.xml:349]: This is a temporary repo for https://github.com/apache/hive/pull/4542

**Comment by LuciferYang** [comment]
Thank you for your efforts on this. @pan3793

**Comment by pan3793** [comment]
cc @sunchao @JoshRosen @HyukjinKwon @dongjoon-hyun @srowen @mridulm, what do you think about this approach?

**Comment by srowen** [comment]
We should probably do something like this for 4.0, yes.
Why does it require removing the isolated classloader? That's the only part I'm concerned about breaking something else.

**Comment by pan3793** [comment]
> We should probably do something like this for 4.0, yes.

@srowen yes, it's target to 4.0

> Why does it require removing the isolated classloader?

Guava is marked as shared, which means that Guava is not isolated totally. (I don't understand the background, but the fact is, it's required, excluding it from shared classes breaks the test)

https://github.com/apache/spark/blob/afcccb42c96cc7785a85c9463523f34d7a900a1d/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala#L205C3-L219

So technically, if we want to upgrade Guava with `IsolatedClientLoader`, we should make all Hive versions compatible with the new Guava, it's impossible.

And as I comment in https://www.mail-archive.com/dev@spark.apache.org/msg30708.html, I think the `IsolatedClientLoader` is not required after dropping Java 8

> Dropping support for Java 8 means dropping support for Hive lower than 
2.0(exclusive)[1].
> 
> IsolatedClientLoader aims to allow using different Hive jars to communicate 
with different versions of HMS. AFAIK, the current built-in Hive 2.3.9 client 
works well on communicating with the Hive Metastore server through 2.1 to 3.1 
(maybe 2.0 too, not sure). This brings a new question, does the 
IsolatedClientLoader required then?
> 
> I think we should drop IsolatedClientLoader because
> 
> 1. As explained above, we can use the built-in Hive 2.3.9 client to communicate 
with HMS 2.1+
> 2. Since SPARK-42539[2], the default Hive 2.3.9 client does not use 
IsolatedClientLoader, and as explained in SPARK-42539, IsolatedClientLoader 
causes some inconsistent behaviors.
> 3. It blocks Guava upgrading. HIVE-27560[3] aim to make Hive 2.3.10(unreleased) 
compatible with all Guava 14+ versions, but unfortunately, Guava is marked as 
`isSharedClass`[3] in IsolatedClientLoader, so technically, if we want to 
upgrade Guava we need to make all supported Hive versions(through 2.1.x to 
3.1.x) to support high version of Guava, I think it's impossible.
> 
> [1] sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientVersions.scala
> [2] https://issues.apache.org/jira/browse/SPARK-42539
> [3] https://issues.apache.org/jira/browse/HIVE-27560
> [4] https://github.com/apache/spark/pull/33989#issuecomment-926277286

**Comment by sunchao** [comment]
Dropping `IsolatedClientLoader` is also my main concern here - I'm not sure whether anyone in the community still rely on the feature for some reason (cc @xkrogen). I do agree that Hive 2.3.9/10 should be compatible with _most_ Hive versions.

Perhaps it's worth checking out in the dev mailing list, if it's not done already.

**Comment by pan3793** [comment]
> Perhaps it's worth checking out in the dev mailing list, if it's not done already.

@sunchao I posted the proposal in the mailing list thread [What else could be removed in Spark 4?](https://www.mail-archive.com/dev@spark.apache.org/msg30708.html).

**Comment by pan3793** [comment]
> I do agree that Hive 2.3.9/10 should be compatible with most Hive versions.

Yes, I can contribute some use cases.

- We use Hive 2.3.9 to access HMS 2.1/2.3 internally for several years. 
- In default setup of [E-MapReduce of AliCloud](https://www.alibabacloud.com/help/en/e-mapreduce/latest/emr-5-13-x-version-notes), it uses Spark 3.3.1(Hive 2.3.9) to access HMS 3.1.

**Comment by xkrogen** [comment]
> Dropping `IsolatedClientLoader` is also my main concern here - I'm not sure whether anyone in the community still rely on the feature for some reason (cc @xkrogen). I do agree that Hive 2.3.9/10 should be compatible with _most_ Hive versions.

Thanks for the ping @sunchao! We're still using it in Spark 3.1.1 to access Hive 1.1, but we're working to get rid of it after the compatibility work you did on the HMS side (thanks for that!). Our only blocker has been some rollout concerns, nothing from the compatibility angle. So, good from my perspective.

**Comment by JoshRosen** [comment]
Regarding the proposed removal of `IsolatedClientLoader`, I am concerned that doing so might break some users' ability to interface with very old Hive metastores:

Although newer Hive client versions are compatible with a wider range of metastore versions, I am unsure whether they cover _all_ old versions. If users are relying on Hive 0.13, for example, then I am concerned that removing `IsolatedClientLoader` might prevent those users from being able to upgrade Spark without first upgrading their metastore.

As was pointed out upthread in https://github.com/apache/spark/pull/42493#issuecomment-1678857243, though, those old Hive versions require Java 8. If Spark 4.0 drops JDK8 support then we'd already lose support for Hive < 2.0 using out-of-the-box Hive metastore clients.

However, some users run forked versions of old Hive which _do_ support newer JDKs and for them the `IsolatedClientLoader` removal would create pains for their Spark 4.0 upgrade. Since those users are _already_ forking Hive, though, I suppose they could simply recompile their Hive fork against Spark's newer Guava version.

---

That said, I wonder whether we can decouple the `IsolatedClientLoader` removal decision from the Guava upgrade by removing Guava from `sharedClasses` (previously discussed at  https://github.com/apache/spark/pull/33989#issuecomment-928616327). This could also prove useful in case we uncover some other reason why we might want to continue using `IsolatedClientLoader`.

I think Guava would only need to be shared if we were passing instances of Guava classes across the classloader boundary, but I don't think we're doing that.

I tried removing Guava from `sharedClasses` and used SBT to run `org.apache.spark.sql.hive.client.HiveClientSuites` and tests seem to pass. Given this, maybe we could try kicking off a full CI run to see if we get any failures there? I think it may be worth re-testing here in order to verify our assumptions about Guava and the `IsolatedClientLoader`.

**Comment by pan3793** [comment]
@JoshRosen thanks for the comments, https://github.com/apache/spark/pull/42599 is opened to try removing Guava from `sharedClasses`, let's wait for CI results.

> If Spark 4.0 drops JDK8 support then we'd already lose support for Hive < 2.0 using out-of-the-box Hive metastore clients.

Yes, definitely, from the community perspective, that means Spark 4.0 drops support for HMS < 2.0.

> However, some users run forked versions of old Hive which do support newer JDKs and for them the `IsolatedClientLoader` removal would create pains for their Spark 4.0 upgrade. Since those users are already forking Hive, though, I suppose they could simply recompile their Hive fork against Spark's newer Guava version.

I think it's out of the scope of the Spark community's responsibility. And it's very weird to combine Hive 0.13(released in 2014) with Spark 4.0(will release in 2024), considering that Spark 4.0 decided to drop support for Java 8 and Hadoop 2, I don't think we should insist to support such an old Hive version.

I definitely see lots of users fork and maintain their internal Hadoop/Hive/Spark versions, but as you said, they could simply recompile their Hive fork against Spark's newer Guava version, I suppose they also could recompile their Spark fork against their Hive fork. 

In short, I think it does not make sense that a user supposes Spark 4.0.0 should work out-of-box with a forked Hive released a decade ago. He/she should expect some code changes on Spark to make such a combination work.

**Comment by pan3793** [comment]
UPDATE

Dropping `IsolatedClientLoader` is not necessary after we dropped support in SPARK-45328, see technical details at https://github.com/apache/spark/pull/42599.

I think there are no known technical issues blocking us from upgrading Guava while keeping IsolatedClientLoader in Spark 4.0.0

Summary of upgrading Guava steps:

1. SPARK-45292: Remove Guava from shared classes from IsolatedClientLoader https://github.com/apache/spark/pull/42599
2. Wait for Hive 2.3.10 for HIVE-27560
3. Upgrade Spark built-in Guava to 32+

**Comment by dongjoon-hyun** [comment]
Thank you for updating this, @pan3793 .

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala [hunk 2]
```diff
@@ -341,7 +341,7 @@ class IntervalExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
 
     Seq(
       (Period.ofMonths(2), Int.MaxValue) -> "overflow",
-      (Period.ofMonths(Int.MinValue), 10d) -> "not in range",
+      (Period.ofMonths(Int.MinValue), 10d) -> "out of range",
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala:344]: So, this error message is originated from Guava?

File: dev/test-dependencies.sh [hunk 3]
```diff
@@ -49,7 +49,7 @@ OLD_VERSION=$($MVN -q \
     --non-recursive \
     org.codehaus.mojo:exec-maven-plugin:1.6.0:exec | grep -E '[0-9]+\.[0-9]+\.[0-9]+')
 # dependency:get for guava and jetty-io are workaround for SPARK-37302.
-GUAVA_VERSION=$(build/mvn help:evaluate -Dexpression=guava.version -q -DforceStdout | grep -E "^[0-9.]+$")
+GUAVA_VERSION=$(build/mvn help:evaluate -Dexpression=guava.version -q -DforceStdout | grep -E "^[0-9\.]+")
```
Comment (pan3793) [inline dev/test-dependencies.sh:52]: update because the new Guava versions have `-jre` or `-android` suffix

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala (hunk 2 above)
Comment (pan3793) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala:344]: Yes

File: pom.xml [hunk 4]
```diff
@@ -3446,6 +3446,7 @@
             <includes>
               <include>org.spark-project.spark:unused</include>
               <include>com.google.guava:guava</include>
+              <include>com.google.guava:failureaccess</include>
```
Comment (pan3793) [inline pom.xml:3449]: I refer to the shading rules in the connect module to handle Guava dependencies, cc @LuciferYang, would be great if you have an opportunity to take a look.

File: assembly/pom.xml [hunk 5]
```diff
@@ -75,11 +75,7 @@
       <version>${project.version}</version>
     </dependency>
 
-    <!--
-      Because we don't shade dependencies anymore, we need to restore Guava to compile scope so
-      that the libraries Spark depend on have it available. We'll package the version that Spark
-      uses (14.0.1) which is not the same as Hadoop dependencies, but works.
-    -->
+    <!-- Restore Guava to compile scope so that the libraries Spark depend on have it available -->
```
Comment (pan3793) [inline assembly/pom.xml:78]: the existing comments are outdated, Hadoop already switched to `hadoop-shaded-guava-1.2.0.jar`

Comment (pan3793) [reply assembly/pom.xml:78]: One additional question is should we remove `<scope>${hadoop.deps.scope}</scope>`?

File: dev/test-dependencies.sh (hunk 3 above)
Comment (dongjoon-hyun) [reply dev/test-dependencies.sh:52]: Please open a new PR for this, @pan3793 . It's worth to have an independent JIRA issue.

File: assembly/pom.xml (hunk 5 above)
Comment (dongjoon-hyun) [reply assembly/pom.xml:78]: Please open a new JIRA and PR to discuss this.

File: pom.xml (hunk 4 above)
Comment (LuciferYang) [reply pom.xml:3449]: fine to me

File: dev/test-dependencies.sh (hunk 3 above)
Comment (pan3793) [reply dev/test-dependencies.sh:52]: opened SPARK-48254

File: assembly/pom.xml (hunk 5 above)
Comment (pan3793) [reply assembly/pom.xml:78]: opened SPARK-48255

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala (hunk 2 above)
Comment (pan3793) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala:344]: cc @cloud-fan do you think such error message change is acceptable?

**Comment by pan3793** [comment]
@JoshRosen @srowen @sunchao @dongjoon-hyun @LuciferYang @HyukjinKwon I think we are ready to upgrade Guava to the modern version, could you please take a look?

**Comment by LuciferYang** [comment]
should we test 33.2.0?

**Comment by pan3793** [comment]
> should we test 33.2.0?

`33.1.0-jre` is chosen because Spark uses `33.1.0-jre` in other places, and after this change, we can unify all Guava versions.

**Comment by LuciferYang** [comment]
> > should we test 33.2.0?
> 
> `33.1.0-jre` is chosen because Spark uses `33.1.0-jre` in other places, and after this change, we can unify all Guava versions.

OK

**Comment by LuciferYang** [comment]
Is there a possibility that upgrading Guava could carry similar risks as described in https://github.com/apache/spark/pull/46528 ? Could there be legacy Hive UDF jars that are bound to Guava 14.0.1?

WDYT @dongjoon-hyun ?

**Comment by LuciferYang** [comment]
Should we continue to push this one forward？

File: dev/deps/spark-deps-hadoop-3-hive-2.3 [hunk 6]
```diff
@@ -63,12 +63,14 @@ derby/10.16.1.1//derby-10.16.1.1.jar
...
 dropwizard-metrics-hadoop-metrics2-reporter/0.1.2//dropwizard-metrics-hadoop-metrics2-reporter-0.1.2.jar
+error_prone_annotations/2.26.1//error_prone_annotations-2.26.1.jar
 esdk-obs-java/3.20.4.2//esdk-obs-java-3.20.4.2.jar
+failureaccess/1.0.2//failureaccess-1.0.2.jar
 flatbuffers-java/23.5.26//flatbuffers-java-23.5.26.jar
 gcs-connector/hadoop3-2.2.21/shaded/gcs-connector-hadoop3-2.2.21-shaded.jar
 gmetric4j/1.0.10//gmetric4j-1.0.10.jar
 gson/2.2.4//gson-2.2.4.jar
-guava/14.0.1//guava-14.0.1.jar
+guava/33.1.0-jre//guava-33.1.0-jre.jar
```
Comment (LuciferYang) [inline dev/deps/spark-deps-hadoop-3-hive-2.3:73]: we can upgrade to test 33.2.1 now

**Review by dongjoon-hyun — COMMENTED** [review]
Could you rebase this PR once more, @pan3793 ?

I'm wondering if we want to deliver this via Apache Spark 4.0.0-preview2 next Monday.

**Comment by pan3793** [comment]
@dongjoon-hyun I rebased this PR and updated the description. Guava upgrading was requested many times from various users, it would be great if we could deliver this Guava upgradig in the coming Spark 4.0.0-preview2

**Comment by dongjoon-hyun** [comment]
I agree with you, @pan3793 .

**Review by dongjoon-hyun — COMMENTED** [review]
~BTW, the latest one is `33.3.0-jre`. Could you update this PR if there is no reason to stick to old one~, @pan3793 ?
- https://mvnrepository.com/artifact/com.google.guava/guava/33.3.0-jre

Never mind. I saw that it's consistent with the existing Guava version.

https://github.com/apache/spark/blob/e037953d4c5879f927c527685c2d027fe2fc08c1/pom.xml#L294

File: dev/deps/spark-deps-hadoop-3-hive-2.3 [hunk 7]
```diff
@@ -33,6 +33,7 @@ breeze-macros_2.13/2.1.0//breeze-macros_2.13-2.1.0.jar
 breeze_2.13/2.1.0//breeze_2.13-2.1.0.jar
 bundle/2.24.6//bundle-2.24.6.jar
 cats-kernel_2.13/2.8.0//cats-kernel_2.13-2.8.0.jar
+checker-qual/3.42.0//checker-qual-3.42.0.jar
```
Comment (dongjoon-hyun) [inline dev/deps/spark-deps-hadoop-3-hive-2.3:36]: I guess we need to update `LICENSE-binary` and `NOTICE-binary` for new jars, @pan3793 .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0-preview2.

**Comment by dongjoon-hyun** [comment]
Thank you, @pan3793 and @LuciferYang .

I want to ping you guys for head-ups once more, @JoshRosen , @cloud-fan , @HyukjinKwon , @yaooqinn , @mridulm , @viirya , @huaxingao .

File: dev/deps/spark-deps-hadoop-3-hive-2.3 (hunk 7 above)
Comment (dongjoon-hyun) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:36]: Please proceed as a follow-up, @pan3793 .
